@here/harp-mapview/lib/CameraUtils.ts

```diff
@@ -4,33 +4,128 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+import { Vector2Like } from "@here/harp-geoutils";
+import { assert } from "@here/harp-utils";
 import * as THREE from "three";
 
 import { MAX_FOV_RAD, MIN_FOV_RAD } from "./FovCalculation";
 
+function isCenteredProjection(ppalPoint: Vector2Like): boolean {
+    return ppalPoint.x === 0 && ppalPoint.y === 0;
+}
+
+function computePosSideFov(focalLength: number, ppOffset: number, viewportSide: number): number {
+    return Math.atan(((1 - ppOffset) * viewportSide * 0.5) / focalLength);
+}
+
+function computeNegSideFov(focalLength: number, ppOffset: number, viewportSide: number): number {
+    return Math.atan(((1 + ppOffset) * viewportSide * 0.5) / focalLength);
+}
+
+function setFovs(
+    camera: THREE.PerspectiveCamera,
+    ppalPoint: Vector2Like,
+    focalLength: number,
+    height: number,
+    hFov: number
+): void {
+    if (isCenteredProjection(ppalPoint)) {
+        delete camera.userData.fovs;
+    } else {
+        const width = height * camera.aspect;
+        camera.userData.fovs = {
+            top: computePosSideFov(focalLength, ppalPoint.y, height),
+            right: computePosSideFov(focalLength, ppalPoint.x, width),
+            horizontal: hFov
+        };
+    }
+}
+
+function setFocalLength(camera: THREE.PerspectiveCamera, focalLength: number): void {
+    camera.userData.focalLength = focalLength;
+}
+
+function vertToHorizSymFov(camera: THREE.PerspectiveCamera): number {
+    return 2 * Math.atan(Math.tan(THREE.MathUtils.degToRad(camera.fov) / 2) * camera.aspect);
+}
+
+/**
+ * Computes a camera's focal length from vertical fov and viewport height or horizontal fov and
+ * viewport width.
+ * @beta
+ *
+ * @param fov - Vertical or horizontal field of view in radians.
+ * @param viewportSide - Viewport height if fov is vertical, otherwise viewport width.
+ * @param ppOffset - Principal point offset in y direction if fov is vertical,
+ * otherwise in x direction.
+ * @returns focal length in pixels.
+ */
+function computeFocalLength(fov: number, viewportSide: number, ppOffset: number): number {
+    const halfSide = viewportSide / 2;
+    const ppCentered = ppOffset === 0;
+    if (ppCentered) {
+        return halfSide / Math.tan(fov / 2);
+    }
+
+    const ppOffsetSq = ppOffset ** 2;
+
+    // Special case for (close to) right angle fov, tangent approaches infinity.
+    const eps = 1e-6;
+    if (Math.abs(fov - Math.PI / 2) < eps) {
+        // docs
+        return halfSide * Math.sqrt(1 - ppOffsetSq);
```

Comment:
If the angle is close to a right angle, won't the focal length be (almost) 0?

Reply:
I've added quite some documentation explaining everything, hope it's clearer now.

Reply:
Thanks, I was confused, now it makes sense.